src/sparse/merkle_tree.rs

```diff
@@ -0,0 +1,445 @@
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "108f731f2414e33ae57e584dc26bd276db07874436b2264ca6e520c658185c6b";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_100() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..100 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let root = tree.root();
+        let expected_root = "82bf747d455a55e2f7044a03536fc43f1f55d43b855e72c0110c986707a23e4d";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_empty() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        let key = 0_u32.to_be_bytes();
+        let data = [0_u8; 0];
+        tree.update(&key, &data).unwrap();
+
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_1_delete_1() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        let key = 1_u32.to_be_bytes();
+        let data = b"DATA";
+        tree.update(&key, data).unwrap();
+        tree.delete(&key).unwrap();
+
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_1_update_empty() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        let key = 0_u32.to_be_bytes();
+        let data = b"DATA";
+        tree.update(&key, data).unwrap();
+        tree.update(&key, &[0; 0]).unwrap();
+
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_2_delete_1() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage);
+
+        for i in 0_u32..2 {
+            let key = i.to_be_bytes();
+            let data = b"DATA";
+            tree.update(&key, data).unwrap();
+        }
+
+        let key = 0_u32.to_be_bytes();
+        tree.delete(&key).unwrap();
+
+        let root = tree.root();
+        let expected_root = "d7cb6616832899ac111a852ca8df2d63a1cdb36cb84651ffde72e264506a456f";
```

Comment:
This doesn't match `0x39f36a7cb4dfb1b46f03d044265df6a491dffc1034121bc1071a34ddce9bb14b` from the specs

Reply:
Updated